src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DesignTimeBuilds/DesignTimeBuildLoggerProvider.cs

```diff
@@ -0,0 +1,209 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.Build.Framework;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.ProjectSystem.Build;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.ProjectSystem.VS.UI.InfoBarService;
+using Microsoft.VisualStudio.Telemetry;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.DesignTimeBuilds;
+
+/// <summary>
+/// An MSBuild logger provider, that observes design-time builds for diagnostic purposes.
+/// </summary>
+/// <remarks>
+/// This logger is only active during design-time builds. It captures the most time-consuming targets
+/// and reports them via telemetry. It also captures data about failed DTBs.
+/// </remarks>
+[method: ImportingConstructor]
+[Export(typeof(IBuildLoggerProviderAsync))]
+[AppliesTo(ProjectCapability.DotNet)]
+internal sealed class DesignTimeBuildLoggerProvider(ITelemetryService telemetryService, IInfoBarService infoBarService, IProjectFaultHandlerService projectFaultHandlerService, ConfiguredProject project) : IBuildLoggerProviderAsync
+{
+    public Task<IImmutableSet<ILogger>> GetLoggersAsync(IReadOnlyList<string> targets, IImmutableDictionary<string, string> properties, CancellationToken cancellationToken)
+    {
+        IImmutableSet<ILogger> loggers = ImmutableHashSet<ILogger>.Empty;
+
+        if (properties.GetBoolProperty("DesignTimeBuild") is true)
+        {
+            loggers = loggers.Add(new DesignTimeTelemetryLogger(telemetryService, infoBarService, projectFaultHandlerService, project));
+        }
+
+        return Task.FromResult(loggers);
+    }
+
+    private sealed class DesignTimeTelemetryLogger(ITelemetryService telemetryService, IInfoBarService infoBarService, IProjectFaultHandlerService projectFaultHandlerService, ConfiguredProject project) : ILogger
+    {
+        /// <summary>
+        /// Stores hashed target names so that we don't recompute these hashes for every project configuration.
+        /// The set of target names is relatively small.
+        /// </summary>
+        private static ImmutableDictionary<(string Target, bool IsMicrosoft), string> s_hashByTargetName = ImmutableDictionary<(string Target, bool IsMicrosoft), string>.Empty;
```

Comment:
nit: is the bool `IsMicrosoft` refer to internal users? if so, can we rename it to `IsInternalUser` or something like that?

Reply:
or does it refer to a Microsoft's target?

Reply:
After reading all your changes, I think it makes sense now :)